docs/public-api.md

```diff
@@ -0,0 +1,101 @@
+
+##### What's the point of this restriction?
+
+NIO in a later version may conform its type to said protocol, or the package that provides the protocol may conform the NIO type to _their_ protocol.
+
+##### Examples
+
+  - ✅ `extension EventLoopFuture: MyOwnProtocol { ... }`, assuming `MyOwnProtocol` lives in your codebase
+  - ❌ `extension EventLoopFuture: DebugStringConvertible { ... }`, `DebugStringConvertible` is a standard library protocol
+
+---
+
+### 3. Conforming your types to NIO protocols
+
+It is acceptable to conform _your own types_ to NIO protocols but it is not acceptable to conform types you do not control to NIO protocols.
+
+##### What's the point of this restriction?
+
+NIO in a later version might add this conformance or the owner of the type may add the conformance.
+
+##### Examples
+
+  - ✅ `extension MyHandler: ChannelHandler { ... }`
+  - ❌ `extension Array: EventLoopGroup where Element: EventLoop { ... }`, as `Array` lives in the standard library
+
+---
+
+### 4. Extending NIO types
+
+It is acceptable to extend NIO types with `public` methods/properties that either use one of your types as a non-default argument or are prefixed in a way that prevents clashes. Adding `private`/`internal` methods/properties is always acceptable.
+
+#### What's the point of these restrictions?
+
+NIO might later add a member function/property with the same signature as yours.
+
+##### Examples
+
+  - ✅ `extension ByteBuffer { public mutating func readMyType(at: Int) -> MyType {...} }`, acceptable because it returns a `MyType`
+  - ✅ `extension ByteBuffer { public mutating func myProjectReadInteger(at: Int) -> Int {...} } `, acceptable because prefixed with `myProject`
+  - ✅ `extension ByteBuffer { internal mutating func readFloat(at: Int) -> Float {...} }`, acceptable even though `Float` is not in your control because the function is `internal`
+  - ❌ `extension ByteBuffer { public mutating func readBool(at: Int) -> Bool {...} }`, because `Bool` is a standard library type not in your control
+
+
+## Promises the NIO team make
+
+Before releasing a new major version, i.e. SwiftNIO 3.0.0 we promise the following:
+
+### 1. No global namespace additions that aren't prefixed
+
+In a minor or patch version, all global types/modules/functions that we will add will have `NIO*`/`nio*`/`CNIO*`/`cnio*`/`c_nio*` prefixes, this includes C symbols that are visible to others. Be aware that we don't control the exact versions of the system libraries installed on your system so we can't make any guarantees about the symbols exported by the system libraries we depend on.
+
+
+##### What does this mean concretely?
+
+- ✅ We might add a new global type called `NIOJetPack`.
+- ✅ We might add a new module called `NIOHelpers`.
+- ✅ We might add a new global functions called `nioRun` (very unlikely) or `c_nio_SSL_write` (more likely).
+- ❌ We will not add a global new type called `Tomato`.
```

Comment:
As discussed with @weissi OOB, I'm not sure that this is necessary. Shadowing should work here too, no? (i.e. the new type doesn't introduce compiler amgiguity?) (this specific section lacks the "What's the point of this restriction?" :-)

This is slightly off-topic, but IMO all stuff should either be prefixed `NIO` or not, for API consistency. When looking at the sidebar of https://apple.github.io/swift-nio/docs/current/NIO/index.html those `NIO` types immediately hit your eye as outliers.
Right now NIO mostly gets away w/o prefixes because it just renamed all types Foundation already has (`EventLoop` instead of `RunLoop` etc, i.e. it used the Java terminology). I would personally either continue doing this (and e.g. event new names for `NIOAny`, maybe `ChannelBucket`?), or prefix everything for consistency and ideally rename stuff to match Foundation (e.g. `NIORunLoop` instead of `EventLoop`).

Reply:
@helje5 let's assume three modules:

1. `NIO` which in version 2.0 does not declare a `struct Tomato`
2. `Caprese` which declares a `public struct Tomato`
3. `MyApp`

Let's also assume that `MyApp` has both `import Caprese` and `import NIO` and it uses `Tomato` (implicitly from `Caprese` for example like this):

```
let t = Tomato()
```

If now in version NIO in version `2.1.0` introduces a `public struct Tomato { ... }` as well, then now compilation of `MyApp` will fail will an ambiguity error.

The only way to resolve it is that NIO introduces it as `NIOTomato` and everything is good. Does that make sense?

Reply:
btw, event loop is also not "Java terminology" it's just what it is. Run loop is less specific.

Reply:
Makes sense, maybe you want to add the explanation to the document (like you did in the other sections).
(P.S.: I wanted to say "Netty terminology", but that is nitpicking. I think the point in itself is valid).

Reply:
@helje5 added a paragraph, useful?